Replace Management Samples (#343)
SteeltoeOSS/Samples: Management/src/ActuatorApi/README.md

### MySQL

This application depends on a MySQL database. Refer to [Common Tasks](../../../CommonTasks.md#MySQL) to start a server that matches the credentials defined in [appsettings.Development.json](./appsettings.Development.json).

### Administrative Tasks (initialize the database)

In order to demonstrate [Steeltoe Management Tasks](https://docs.steeltoe.io/api/v3/management/tasks.html), the database schema and its contents are managed as administrative tasks.

1. Apply Entity Framework Core database schema migration scripts:

    ```shell
    dotnet run --runtask=MigrateDatabase
    ```

1. Run [`ForecastTask`](./AdminTasks/ForecastTask.cs) to predict weather for the next 7 days:

    ```shell
    dotnet run --runtask=ForecastWeather
    ```

    *Optional* - Add forecast data starting from a specific date and/or number of days:

    ```shell
    dotnet run --runtask=ForecastWeather --fromDate=10/18/2024 --days=30
    ```

> [!NOTE]  
> For the `fromDate` parameter, use values formatted as `yyyy-dd-MM` or `MM/dd/yyyy`.

> [!TIP]
> The task [`ResetTask`](./AdminTasks/ResetTask.cs) can be used to remove all forecast data:

> > `dotnet run --runtask=ResetWeather`

Reviewer: This indent is wrong now too.

Reviewer: This is a limitation in GitHub markdown, documented at https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts:

> Alerts cannot be nested within other elements.

1. Run the `cf push` command to deploy from source (you can monitor logs with `cf logs actuator-api-management-sample`)
    * When deploying to Windows, binaries must be built locally before push. Use the following commands instead:

      ```shell
      dotnet publish -r win-x64 --self-contained
      cf push -f manifest-windows.yml -p bin/Release/net8.0/win-x64/publish
      ```

1. Copy the value of `routes` in the output and open in your browser. The app should start and respond to requests, but the database still needs to be configured with the tasks listed in the next section.

> [!NOTE]  
> The provided manifest will create an app named `actuator-api-management-sample` and attempt to bind it to the MySql service `sampleMySqlService`.

### Running Tasks

Depending on the steps taken to push the application to Cloud Foundry, the commands below may require customization (for example, if the application was not published before pushing to a Linux cell, the path for the command might be `./bin/Debug/net8.0/linux-x64/Steeltoe.Samples.ActuatorApi`)

1. Apply Entity Framework Core database migration scripts:

    ```shell

>     cf run-task actuator-api-management-sample --command "./Steeltoe.Samples.ActuatorApi runtask=MigrateDatabase" 

Reviewer: Tried this, but it fails with the following line in the logs:
> 2024-12-19T13:02:06.583+01:00 [APP/TASK/a8d13dcd/0] [ERR] bash: line 1: ./Steeltoe.Samples.ActuatorApi: No such file or directory

Am I doing something wrong?

    ```

1. Run [`ForecastTask`](./AdminTasks/ForecastTask.cs) to predict weather for the next 7 days:

    ```shell
    cf run-task actuator-api-management-sample --command "./Steeltoe.Samples.ActuatorApi runtask=ForecastWeather" 
    ```

> [!TIP]  
> To remove all forecast data, run [`ResetTask`](./AdminTasks/ResetTask.cs):
>
> ```shell
> cf run-task actuator-api-management-sample --command "./Steeltoe.Samples.ActuatorApi runtask=ResetWeather" 

> > ```

Reviewer: This indent is wrong now.

Reviewer: This is a limitation in GitHub markdown, documented at https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts:

> Alerts cannot be nested within other elements.

## Running locally

Be sure to follow the [ActuatorWeb Readme](../ActuatorWeb/README.md) to the point where that application is running
before continuing here.

### MySQL

This application depends on a MySQL database. Refer to [Common Tasks](../../../CommonTasks.md#MySQL) to start a server
that matches the credentials defined in [appsettings.Development.json](./appsettings.Development.json).

### Administrative Tasks (initialize the database)

In order to demonstrate [Steeltoe Management Tasks](https://docs.steeltoe.io/api/v3/management/tasks.html), the database
schema and its contents are managed as administrative tasks.

1. Apply Entity Framework database schema migration scripts:

    ```shell
    dotnet run --runtask=MigrateDatabase
    ```

1. Run [`ForecastTask`](./AdminTasks/ForecastTask.cs) to predict weather for the next 7 days:

    ```shell
    dotnet run --runtask=ForecastWeather
    ```

    *Optional* - Add forecast data starting from specific date and/or number of days:

    ```shell
    dotnet run --runtask=ForecastWeather --fromDate=10/18/2024 --days=30
    ```

    > [!NOTE]  

>     > Use values formatted for invariant culture (`MM/dd/yyyy`) for the `fromDate` parameter.

Author: I think since we're spelling out exactly what is expected it should be easy enough to work with, and there's less code to deal with since it's .NET defaults. I wouldn't block the change in the future, but I don't see the need to do it now

>     > Use values formatted for invariant culture (`MM/dd/yyyy`) for the `fromDate` parameter.

Reviewer: The code doesn't need to change, `yyyy-dd-mm` is already accepted when using the invariant culture,

>     > Use values formatted for invariant culture (`MM/dd/yyyy`) for the `fromDate` parameter.

Author: ```suggestion
    > For the `fromDate` parameter, use values formatted as `yyyy-dd-mm` or `MM/dd/yyyy`.
```
How about this?

    > [!TIP]
    > The task [`ResetTask`](./AdminTasks/ResetTask.cs) can be used to remove all forecast data:
    > `dotnet run --runtask=ResetWeather`

1. Use your preferred IDE or `dotnet run` to start the application.

## Running on Tanzu Platform for Cloud Foundry

1. Create a MySQL service instance in an org/space

   ```shell
   cf target -o your-org -s your-space
   ```

    * When using VMware MySQL for Tanzu Application Service:

      ```shell
      cf create-service p.mysql db-small sampleMySqlService
      ```

    * When using the Cloud Service Broker for Azure:

      ```shell
      cf create-service csb-azure-mysql small sampleMySqlService
      ```

    * When using the Cloud Service Broker for GCP:

      ```shell
      cf create-service csb-google-mysql your-plan sampleMySqlService
      ```

1. Wait for the service to become ready (you can check with `cf services`)

> 1. Run the command `cf push` to deploy from source (you can monitor logs with `cf logs actuator-api-management-sample`)

Reviewer: This is "Run the `cf push` command to" in all other samples, but was changed here to "Run the command `cf push` to". If the original text is wrong, it should be updated in other samples too. If not, I don't think we should deviate here.

    * When deploying to Windows, binaries must be built locally before push. Use the following commands instead:

      ```shell
      dotnet publish -r win-x64 --self-contained
      cf push -f manifest-windows.yml -p bin/Release/net8.0/win-x64/publish
      ```

1. Copy the value of `routes` in the output and open in your browser. The app should start and respond to requests, but the database still needs to be configured with the tasks listed in the next section.

> [!NOTE]  
> The provided manifest will create an app named `actuator-api-management-sample` and attempt to bind to the the app to
> MySql service `sampleMySqlService`.

### Running Tasks

Depending on the steps taken to push the application to Cloud Foundry, the commands below may require customization (for
example, if the application was published before pushing, the path in the command should be shortened to
`./Steeltoe.Samples.ActuatorApi`)

1. Apply Entity Framework database migration scripts:

    ```shell

>     cf run-task actuator-api-management-sample --command "./bin/Debug/net8.0/linux-x64/Steeltoe.Samples.ActuatorApi runtask=MigrateDatabase" 

Author: yeah, in both cases I wrote out the more complex version.. I'll be pushing an update to make it more consistent

﻿# ActuatorApi - a sample API app for Steeltoe Actuators

ASP.NET Core sample app illustrating how to use [Steeltoe Management Endpoints](https://docs.steeltoe.io/api/v3/management/) together with the [Apps Manager](https://docs.vmware.com/en/VMware-Tanzu-Application-Service/6.0/tas-for-vms/manage-apps.html) on Cloud Foundry and [Spring Boot Admin](https://docs.spring-boot-admin.com/) anywhere else.

> This application also illustrates how to have application metrics captured and exported to the [Metrics Registrar](https://docs.vmware.com/en/VMware-Tanzu-Application-Service/6.0/tas-for-vms/metric-registrar-index.html) service so that applications metrics can be viewed in any tool that is able to consume those metrics from the [Cloud Foundry Loggregator](https://github.com/cloudfoundry/loggregator-release).  Several tools exist that can do this, including [App Metrics for VMware Tanzu](https://docs.vmware.com/en/App-Metrics-for-VMware-Tanzu/2.2/app-metrics/GUID-index.html).

Reviewer: ```suggestion
This application also illustrates how to have application metrics captured and exported to the [Metrics Registrar](https://docs.vmware.com/en/VMware-Tanzu-Application-Service/6.0/tas-for-vms/metric-registrar-index.html) service so that application metrics can be viewed in any tool that is able to consume those metrics from the [Cloud Foundry Loggregator](https://github.com/cloudfoundry/loggregator-release).  Several tools exist that can do this, including [App Metrics for VMware Tanzu](https://docs.vmware.com/en/App-Metrics-for-VMware-Tanzu/2.2/app-metrics/GUID-index.html).
```
